raspberry-pi/4/poe-hat.nix

```diff
@@ -0,0 +1,122 @@
+                    #cooling-cells = <2>;
+                    cooling-levels = <0 1 10 100 255>;
+                    status = "okay";
+                  };
+                };
+              };
+
+              fragment@1 {
+                target = <&cpu_thermal>;
+                __overlay__ {
+                  trips {
+                    trip0: trip0 {
+                      temperature = <40000>;
+                      hysteresis = <2000>;
+                      type = "active";
+                    };
+                    trip1: trip1 {
+                      temperature = <45000>;
+                      hysteresis = <2000>;
+                      type = "active";
+                    };
+                    trip2: trip2 {
+                      temperature = <50000>;
+                      hysteresis = <2000>;
+                      type = "active";
+                    };
+                    trip3: trip3 {
+                      temperature = <55000>;
+                      hysteresis = <5000>;
+                      type = "active";
+                    };
+                  };
+                  cooling-maps {
+                    map0 {
+                      trip = <&trip0>;
+                      cooling-device = <&fan0 0 1>;
+                    };
+                    map1 {
+                      trip = <&trip1>;
+                      cooling-device = <&fan0 1 2>;
+                    };
+                    map2 {
+                      trip = <&trip2>;
+                      cooling-device = <&fan0 2 3>;
+                    };
+                    map3 {
+                      trip = <&trip3>;
+                      cooling-device = <&fan0 3 4>;
+                    };
+                  };
+                };
+              };
+
+              fragment@2 {
+                target-path = "/__overrides__";
+                __overlay__ {
+                  poe_fan_temp0 =		<&trip0>,"temperature:0";
+                  poe_fan_temp0_hyst =	<&trip0>,"hysteresis:0";
+                  poe_fan_temp1 =		<&trip1>,"temperature:0";
+                  poe_fan_temp1_hyst =	<&trip1>,"hysteresis:0";
+                  poe_fan_temp2 =		<&trip2>,"temperature:0";
+                  poe_fan_temp2_hyst =	<&trip2>,"hysteresis:0";
+                  poe_fan_temp3 =		<&trip3>,"temperature:0";
+                  poe_fan_temp3_hyst =	<&trip3>,"hysteresis:0";
+                };
+              };
+
+              __overrides__ {
+                poe_fan_temp0 =		<&trip0>,"temperature:0";
+                poe_fan_temp0_hyst =	<&trip0>,"hysteresis:0";
+                poe_fan_temp1 =		<&trip1>,"temperature:0";
+                poe_fan_temp1_hyst =	<&trip1>,"hysteresis:0";
+                poe_fan_temp2 =		<&trip2>,"temperature:0";
+                poe_fan_temp2_hyst =	<&trip2>,"hysteresis:0";
+                poe_fan_temp3 =		<&trip3>,"temperature:0";
+                poe_fan_temp3_hyst =	<&trip3>,"hysteresis:0";
```

Comment:
These are `dtparam` options; e.g. `dtparam=poe_fan_temp0=XXX,poe_fan_temp0_hyst=YYY` etc...

Ideally we'd make them options.

The way this maps is:

```dts
// param_name = <&label>,"property   :offset""
poe_fan_temp0 = <&trip0>,"temperature:0";

trip0: trip0 {
	temperature = <40000>;  <- "0th" item (offset) of property temperature, of node with label trip0
	hysteresis = <2000>;
	type = "active";
};
```

https://www.raspberrypi.org/documentation/configuration/device-tree.mdh

So here it's 4 pairs of params to control the four trip points.

Reply:
Is it safe to change the defaults? Do we want to expose it (with a warning)?

Reply:
No idea.

Only thing I know is that the documentation is here:

  - https://github.com/raspberrypi/linux/blob/bcb79983422d056ea85d2b4a515762c0d56690e5/arch/arm/boot/dts/overlays/README#L2487-L2505

The parameters are not hidden piece of information. But there's no in-depth details either.

Reply:
Yeah, I don't know for sure, but doing some searching I've seen references to tuning the params (e.g. https://www.jeffgeerling.com/blog/2021/taking-control-pi-poe-hats-overly-aggressive-fan ) so it seems like a safe assumption. I'm using this code on a set of four rpi4s and am happy with it, but I can update to include the param options.